visualization/utils/common.py

```diff
@@ -0,0 +1,388 @@
+import datetime
+import functools
+import json
+import logging
+import re
+import subprocess
+from collections import defaultdict
+from typing import Callable, Dict, Iterable, List, Optional, Set, Tuple, Union
+
+from flask import current_app
+from joblib import Parallel, delayed
+from settings.constants import (
+    ANSWER_FIELD,
+    ERROR_MESSAGE_TEMPLATE,
+    OUTPUT,
+    PARAMETERS_FILE_NAME,
+    QUESTION_FIELD,
+    RESULTS_PATH,
+    STATS_KEYS,
+)
+
+from nemo_skills.inference.prompt.few_shot_examples.examples_gsm8k import examples_map as examples_gsm8k
+from nemo_skills.inference.prompt.few_shot_examples.examples_math import examples_map as examples_math
+from nemo_skills.utils import unroll_files
+
+examples = {
+    **examples_gsm8k,
+    **examples_math,
+}
+
+custom_stats = {}
+general_custom_stats = {}
+deleted_stats = set()
+excluded_rows = set()
+
+
+def get_excluded_row() -> Set:
+    return excluded_rows
+
+
+def get_deleted_stats() -> Set:
+    return deleted_stats
+
+
+def get_custom_stats() -> Dict:
+    return custom_stats
+
+
+def get_general_custom_stats() -> Dict:
+    return general_custom_stats
+
+
+def parse_model_answer(answer: str) -> List[Dict]:
+    code_start, code_end = map(

```

Comment:
there are functions in nemo_skills/code_execution/utils.py and nemo_skills/code_execution/math_grader.py that can extract code / errors / answers that you can reuse?

Reply:
I don't quite understand - can you please explain this more?

Reply:
I have the full text as an input and I should extract text, code, and code_output blocks preserving their order. 
extract_error_message from nemo_skills/code_execution/utils.py extracts only last code output block
extract_code_to_execute from nemo_skills/code_execution/utils.py extracts only last code block
extract_answer from nemo_skills/code_execution/math_grader.py extraxtc only value insibe \boxed{}